cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -917,4 +931,105 @@ private boolean deleteIndexRow(Row row) {
   Scanner searchByIndex(String indexColumn, String value) {
     return indexedTable.readByIndex(Bytes.toBytes(indexColumn), Bytes.toBytes(value));
   }
+
+  /**
+   * Starts scanning/deleting of V1 table value row keys in batches. If there are no more value row keys available,
+   * it will scan history rows. If there are no value/history rows, it returns empty list.
+   *
+   * @param limit Batch size for scan/delete.
+   * @param delete Determines if operation to perform is scan or delete.
+   * @return List of {@link KeyValue} where key is timestamp for history row other wise it is null,
+   * value is {@link MetadataEntry}
+   */
+  public List<KeyValue<Long, MetadataEntry>> scanOrDeleteFromV1Table(int limit, boolean delete) {
+    // Try to scan value rows first
+    byte[] rowPrefix = MdsKey.getValueRowPrefix();
+    byte[] stopRowKey = Bytes.stopKeyForPrefix(rowPrefix);
+
+    List<KeyValue<Long, MetadataEntry>> entries = scanOrDelete(rowPrefix, stopRowKey, limit, delete,
+                                                               this::convertFromV1Value);
+
+    // Scan History rows if all the Value rows are scanned
+    if (entries.size() < limit) {
+      rowPrefix = MdsHistoryKey.getHistoryRowPrefix();
+      stopRowKey = Bytes.stopKeyForPrefix(rowPrefix);
+      entries.addAll(scanOrDelete(rowPrefix, stopRowKey, limit - entries.size(), delete, this::convertFromV1History));
+    }
+
+    return entries;
+  }
+
+  private List<KeyValue<Long, MetadataEntry>> scanOrDelete(byte[] rowPrefix, byte[] stopRowKey,
+                                                           int limit, boolean delete,
+                                                           Function<Row, KeyValue<Long, MetadataEntry>> function) {
```

Comment:
Since this method is already taking a `Function`. I think it's better call itself `scanAndProcess` and shouldn't explicitly has a `delete` option, but rather the deletion is done via the `function`. Similarly, this method shouldn't return a `List`, but rather having the `function` does the collection if needed.

Reply:
And potentially, I think you might want to use `Consumer` instead of `Function`.

Reply:
yea, that seems better implementation, thanks for the suggestion!